+per #15279 FSM for PersistentActor

leonidb opened the pull request:
Located in akka-persistence because persistence is experimental and probably should not be used in akka-actor

leonidb committed:
+per #15279 FSM for PersistentActor

ktoso commented:
Hi @leonidb, sorry for the delayed response (most hakkers are currently on vacation skiing or something similar).

I think it's perfectly OK to keep working on this feature in this PR, we can then squash the commits into one or sqaush (git interactive rebase and squash in there, then `push -f`) commits whenever you feel it needs some cleanup. I did not yet get the time to go over your latest work, thanks a look for keeping to work on it! I'll try to have a look soon :-)

leonidb commented:
@ktoso No problem, enjoy your vacation :).
In the mean time I'll try to commit incrementally, so later  it will be easy to reason about the changes one at the time.

patriknw commented on akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala:
```diff
@@ -0,0 +1,178 @@
+      val shoes = Item("2", "Shoes", 89.99F)
+      val coat = Item("3", "Coat", 119.99F)
+
+      within(1.seconds) {
```
I don't think we have to limit this to 1 second. Remove the `within`.

leonidb replied:
I've removed it from both tests

patriknw commented on akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala:
```diff
@@ -0,0 +1,178 @@
+      }
+    }
+  }
+}
```
This is awesome!! Reads very nice.

leonidb replied:
Thanks!

patriknw commented on akka-persistence/src/main/scala/akka/persistence/temp/FSM.scala:
```diff
@@ -0,0 +1,1157 @@
+    /**
+     * Specify domain event that will to be applied when transitioning to the new state.
+     */
+    def applying(domainEvent: E): State[S, D, E] = {
```
I would also like to one variant that takes a callback block as parameter, similar to `persist`. This block should be invoked after the event has been successfully been persisted (and after calling applyState/applyEvent).
That can be used for side effects, e.g. sending a message to someone else.

leonidb replied:
Implemented in my latest commit

patriknw commented:
@leonidb this looks very promising. Great work!

leonidb commented:
@patriknw  thanks!

I've implemented a callback block you've mentioned, here is how it looks in code:
```Scala
case Event(Buy, _)  ⇒ goto(Paid) applying OrderExecuted thenDo { case NonEmptyShoppingCart(items) => reportActor ! PurchaseWasMaid(items)}
```

I've implemented it as a `D ⇒ Unit` function, having a state data as an argument feels useful..

After going through a few names for this handler in DSL I've ended up with `thenDo` - looks concise and works well with the rest of the DSL, I think. Although if not used as the last statement, it won't fit in naturally.

Tell me what you think.

leonidb commented:
@ktoso @patriknw 
Added state timeout:
On State change: Save the timeout as `Option[FiniteDuration]`,
On Recovery: Reset the timeout

Here is how the sample FSM from the test looks like:
```Scala
    startWith(LookingAround, EmptyShoppingCart)

    when(LookingAround) {
      case Event(AddItem(item), _)      ⇒ goto(Shopping) applying ItemAdded(item) forMax(1 seconds)
      case Event(GetCurrentCart, data)  ⇒ stay replying data       //used for assertion of FSM's state during test
    }

    when(Shopping) {
      case Event(AddItem(item), _)      ⇒ stay applying ItemAdded(item) forMax(1 seconds)
      case Event(Buy, _)                ⇒ goto(Paid) applying OrderExecuted thenDo { case NonEmptyShoppingCart(items) => reportActor ! PurchaseWasMaid(items)}
      case Event(Leave, _)              ⇒ stop applying OrderDiscarded thenDo {case _ => reportActor ! ShoppingCardDiscarded}
      case Event(GetCurrentCart, data)  ⇒ stay replying data       //used for assertion of FSM's state during test
      case Event(StateTimeout, _)       ⇒ goto(Inactive) forMax(2 seconds)
    }

    when(Inactive) {
      case Event(AddItem(item), _)      ⇒ goto(Shopping) applying ItemAdded(item) forMax(1 seconds)
      case Event(StateTimeout, _)       ⇒ stop applying OrderDiscarded thenDo {case _ => reportActor ! ShoppingCardDiscarded}
    }

    when(Paid) {
      case Event(Leave, _)              ⇒ stop()
      case Event(GetCurrentCart, data)  ⇒ stay replying data       //used for assertion of FSM's state during test
    }
```

patriknw commented:
state timeout and callback looks good

patriknw commented:
+1 for `andThen`

leonidb commented:
>how about andThen, like in futures?

Agreed, I've committed the change

leonidb commented:
@ktoso @patriknw
Added Java support - uncommented `AbstractFSM` and made it work with the new type

Tell me what you think

leonidb commented:
@ktoso 
I don't know why the build has failed, I ran the build locally with SBT, it passes. Can you have a look? 
Thanks

leonidb commented:
@ktoso @patriknw 
To sum up what I've implemented so far:
1. Copied the `FSM` trait to a temp directory in akka-persistence
2. Implemented persistence of state name, domain events and state timers in `PersistentFSM` trait. 
3. Java support

It looks like I've implemented most of the features we've discussed.
More tests should be added and maybe there are some small bugs here and there, but before I continue I'd like to discuss with you what my next steps should be.

Can you have a look at the whole feature and discuss what's missing, what should be fixed and how can we proceed with this feature.

michaelom commented:
@leonidb 
"Is this an important issue in general or am I mostly doing it for myself?"

This is definitely a big feature for anyone using Akka Persistence for event sourcing, particularly if your aggregate (Partik called it "current state data") comes with some sort of lifecycle.

patriknw commented:
@leonidb thanks for the update. I agree that we should review the full feature and try to identify what is missing, if anything.

One thing you can start thinking about right away is how we should reduce the code duplication between actor.FSM and your clone.

leonidb commented:
@michaelom thanks, hope it will be ready soon

leonidb commented:
@patriknw 
>One thing you can start thinking about right away is how we should reduce the code duplication between actor.FSM and your clone.

I'll look at it

patriknw commented:
Another thing is protobuf serialization. You can find inspiration here https://github.com/akka/akka/blob/master/akka-persistence/src/main/scala/akka/persistence/serialization/MessageSerializer.scala#L109 of how to include a user payload, e.g. for the domainEvent in the StateChangeEvent.

The State in the StateChangeEvent should be serialized as something simple, as a String, so that the user doesn't have to write their own serializer for that.

leonidb commented:
@patriknw 
> You can find inspiration here...

Thanks, I'll look at the example

leonidb commented:
@ktoso @patriknw 
My teammates have started using `PersistentFSMActor` recently and I'm starting to get the feedback. 

While everybody is excited about persistence and recovery of state and state data, I've received some complaints regarding state timeouts.
The problem is the timeouts set with `forMax` are reset with each message, both when state is changed and when staying in the same state, even when the message is unhandled by the FSM, this makes timeouts very unpredictable.

It would be useful to have a way to set a "State Transition Timeout" for staying in a certain state, until FSM transition to the next state. "State Transition Timeout" configuration can be stored in a `Map[S, Duration]`.

Timeouts are implemented in `FSM` trait, there are 3 different ways to configure/affect timeouts. One of those three ways is in `when(stateName: S, stateTimeout: FiniteDuration = null)`, it's stored in 
```Scala
private val stateTimeouts = mutable.Map[S, Timeout]()
```
but this timeout behaves differently, it serves as a default timeout until the next received message.

Are there any plans to change the way timeouts work in `FSM`?  For example, to make `stateTimeouts` behave the way I described.

Do you think "State Transition Timeout" functionality can be useful in `FSM`?

patriknw commented:
@leonidb sorry for the delay

I'm not sure I understand what you mean with "State Transition Timeout". Would it automatically go to another state after this time?

Can't that be implemented less magically with a [timer](http://doc.akka.io/docs/akka/snapshot/scala/fsm.html#Timers)?

patriknw commented on akka-persistence/src/main/scala/akka/persistence/PersistentFSMActor.scala:
```diff
@@ -0,0 +1,113 @@
+   * @param state FSM state
+   * @param domainEvent FSM domain event
+   */
+  private case class StateChangeEvent[S, E](state: S, domainEvent: Option[E], timeout: Option[FiniteDuration]) extends PersistentFsmEvent
```
How will we serialize `S`?
Java serialization is to dangerous, since S can be anything.
It would be too inconvenient for the user to mandate custom serializer for each state.

One idea is to demand that `S` is a subclass of:

    trait FSMState {
        def identifier: String
    }

Then we can serialize only the identifier, and maintain a mapping between identifier and corresponding `S`.

WDYT?

leonidb replied:
What if a user define a state with additional data:
```Scala
  case class SomeState(id: String, someData: String) extends FSMState {   
    def identifier = id    
  }
```
Can we recommend the user in documentation not to add additional state to FSMState implementation, or else he'll have to implement a custom serializer?

Another approach can be to define FSMState as final:
```Scala
case class FSMState(identifier: String)
```
Effectively it's the same as restricting a State type to a String.

patriknw commented on akka-persistence/src/main/scala/akka/persistence/PersistentFSMActor.scala:
```diff
@@ -0,0 +1,113 @@
+ * This is an EXPERIMENTAL feature and is subject to change until it has received more real world testing.
+ */
+abstract class AbstractPersistentFsmActor[S,D,E] extends AbstractFSM[S,D,E] with PersistentFsmActor[S,D,E] {
+  import java.util.function.Consumer
```
is this java8? we are still on java6, but we will move to java8 soon, so this might be ok

hepin1989 replied:
it's java8.

leonidb replied:
@patriknw 
`AbstractFSM` is already implemented this way:
```Scala
/**
 * Java API: compatible with lambda expressions
 *
 * Finite State Machine actor abstract base class.
 *
 * This is an EXPERIMENTAL feature and is subject to change until it has received more real world testing.
 */
abstract class AbstractFSM[S, D, E] extends FSM[S, D, E] {
```
I've copied the comment from it.

patriknw commented:
@leonidb I took a fresh look at this. If we can solve the serialization and reduce the duplication with the original FSM I think this is a great scope for merging, and let others try it out.

leonidb commented:
@patriknw 
>I'm not sure I understand what you mean with "State Transition Timeout". Would it automatically go to another state after this time?

No, FSM Actor would receive some kind of  StateTimeout message. By "State Transition Timeout" I mean setting a "deadline" on staying in a state, regardless of how many messages where received during staying in the state. This kind of timeout seems useful - at least we use it.

>Can't that be implemented less magically with a timer?

Yes, it can. That's how we implemented it. Since timeout set by `timer` is not persisted, in `PersistentFSM` it should be set in 2 places in code - on transition to the state and on recovery, based on the recovered state. Not difficult to implement, but would be more convenient to have it as part of DSL.

leonidb commented:
@patriknw 
>If we can solve the serialization and reduce the duplication with the original FSM I think this is a great scope for merging, and let others try it out.

Sounds great!
I've already started working on implementing protobuf serialization, haven't looked at reducing duplication yet.

rkuhn commented on akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala:
```diff
@@ -0,0 +1,339 @@
+  case object OrderDiscarded extends DomainEvent
+
+  //Side effects - report events to be sent to some "Report Actor"
+  sealed trait ReportEvent
```
nothing extends this trait so far

leonidb replied:
The classes defined in the next lines were supposed to extend it. Fixed.

rkuhn commented on akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala:
```diff
@@ -0,0 +1,339 @@
+
+  //Side effects - report events to be sent to some "Report Actor"
+  sealed trait ReportEvent
+  case class PurchaseWasMaid(items: Seq[Item])
```
`s/Maid/Made/`

leonidb replied:
Fixed

rkuhn commented on akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala:
```diff
@@ -0,0 +1,339 @@
+      case Event(GetCurrentCart, data)  ⇒ stay replying data       //used for assertion of FSM's state during test
+    }
+
+    initialize()
```
this should not be necessary, right? it should be done automatically after recovery finishes

leonidb replied:
Right, removed.

rkuhn commented on akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala:
```diff
@@ -0,0 +1,339 @@
+      case Event(AddItem(item), _)      ⇒ stay applying ItemAdded(item) forMax(1 seconds)
+      case Event(Buy, _)                ⇒ goto(Paid) applying OrderExecuted andThen { case NonEmptyShoppingCart(items) => reportActor ! PurchaseWasMaid(items) }
+      case Event(Leave, _)              ⇒ stop applying OrderDiscarded andThen { case _ => reportActor ! ShoppingCardDiscarded }
+      case Event(GetCurrentCart, data)  ⇒ stay replying data       //used for assertion of FSM's state during test
```
This will deactivate the timeout; it is probably better to add the timeout to the state instead of to the transitions.

leonidb replied:
I specifically wanted to test forMax, as it's part of FSM.State DSL. GetCurrentCart was needed in order to check the state of the FSM in some tests.
I don't use GetCurrentCart in timeout tests, to not affect the timeout - not an elegant solution, I know. This would be easier to implement if the timeouts would behave the way I've described.

rkuhn commented on akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala:
```diff
@@ -0,0 +1,339 @@
+    initialize()
+
+    /**
+     * Override this handler to define the action on Domain Event during recovery
```
This is used also during normal operation, right?

leonidb replied:
Yes, fixed the comment.

rkuhn commented on akka-persistence/src/test/scala/akka/persistence/PersistentFSMActorSpec.scala:
```diff
@@ -0,0 +1,339 @@
+  //Commands
+  sealed trait Command
+  case class AddItem(item: Item) extends Command
+  case object Buy extends Command
```
To make these also work for Java we usually do the following:
~~~scala
sealed abstract class Buy extends Command
case object Buy extends Buy
~~~

patriknw replied:
right, but this is a sample/test that is intended for scala consumption, so I don't see the point of adding boilerplate to it

rkuhn replied:
I may be missing something, but I did see these objects and `something$` classes mentioned in the Java files above.

patriknw replied:
ok, I missed that
later there should be a doc sample, and I was thinking this is that one, and then it should be implemented in scala and java standalone

leonidb replied:
So, should I fix this to make it work better with Java, or should it be split later to 2 separate classes anyway?

rkuhn replied:
Thinking about it that way Patrik is right: while it is nice to test that the Scala and the Java version implement the same semantics, we’ll still need to present the full Java docs. So splitting it up by language is the way to go.

rkuhn commented on akka-persistence/src/main/scala/akka/persistence/temp/FSM.scala:
```diff
@@ -0,0 +1,1172 @@
+   * accumulated while processing the last message, possibly domain event and handler
+   * to be executed after FSM moves to the new state (also triggered when staying in the same state)
+   */
+  final case class State[S, D, E](stateName: S, stateData: D, timeout: Option[FiniteDuration] = None, stopReason: Option[Reason] = None, replies: List[Any] = Nil, domainEvent: Option[E] = None, afterTransitionDo: D ⇒ Unit = {_: D ⇒ })(private[akka] val notifies: Boolean = true) {
```
please break up into multiple lines

leonidb replied:
Done.

rkuhn commented:
@leonidb I just now got a large enough time slot to digest your PR, and I really like what I see! I agree that timeout handling in general has room for improvement, the FSM trait’s StateTimeout is not really a state timeout but as you note more of a state-based ReceiveTimeout. While redoing the whole API for persistence anyway, I think this is the right time to introduce changes to how things work:

  * stateTimeout should fire if the state has not changed after its duration has elapsed, regardless of what else happened in between
  * `forMax` should set the stateTimeout with exactly those semantics as well
  * timeouts should be persisted using two data items: the `System.currentTimeMillis` when the state change occurred (or the timeout was set, which can be afterwards) plus the FiniteDuration of the timeout
  * during recovery we’ll have to replay “virtual time”, meaning that we expire timeouts but don’t deliver their messages—that should have been done and led to events prior to the crash/shutdown; the most previously set timeout for the state that was reached after replay should be effective, including that it should fire if its deadline has elapsed—even if that happened while the actor was down.

rkuhn commented:
Ah, I forgot the second piece of feedback: it should be possible to emit multiple domain events for a single transition, just change `Option[E]` to `immutable.Seq[E]` (and provide varargs convenience methods):
~~~scala
goto(NextState) applying (eventA, eventB, eventC) andThen (_ => sender() ! Done)
~~~

leonidb commented:
@rkuhn @patriknw 
> it should be possible to emit multiple domain events for a single transition, just change Option[E] to immutable.Seq[E]

Should those multiple domain events be persisted as a single `StateChangeEvent`?

rkuhn commented:
Ah, this is a core question: the business domain events are what matters, and they should be persisted to the journal so that they can be consumed by others (Actors or whomever). I didn’t quite realize that your current proposal always wraps (and thereby hides) the domain events in a “technical details”—the fact that this PersistentActor is implemented as an FSM. Shouldn’t the state changes be a result of applying the events instead of an extra piece of information? If the state change is persisted then it should also make sense in the business domain, meaning it should be modeled.

I’m not sure yet what this really means—I just hopped across eight time zones—but I think this is what event sourcing gurus would say.

leonidb commented:
@rkuhn 
> Shouldn’t the state changes be a result of applying the events instead of an extra piece of information? If the state change is persisted then it should also make sense in the business domain, meaning it should be modeled.

Current implementation considers incoming messages as commands and FSM states as something relevant to dealing with commands, but not domain events.
State data is built based on domain events, as it was originally proposed by @patriknw , I think.
So currently it's divided:
Commands: FSM state + incoming messages
Domain Events: FSM state data + domain events

> Shouldn’t the state changes be a result of applying the events instead of an extra piece of information? If the state change is persisted then it should also make sense in the business domain, meaning it should be modeled.

`akka.actor.FSM` is very useful for dealing with incoming commands, as an alternative to using `context.become`. When used this way, FSM state changes cannot be triggered by anything other than commands. It's useful for such an FSM to be able to recover it's state from persistence, this was the motivation for this feature, if I'm not mistaken.
FSM with state change triggered by domain events is a different approach - it can be implemented by persistent Actor with some kind of internal FSM implementation (probably not related to `akka.actor.FSM` as this internal FSM doesn't need to be an `Actor` at all).

The second way can be easily fit into event sourcing, but the first can be more useful.
I can't think of a natural way to combine both approaches.
@rkuhn @patriknw 
We should discuss this further and decide what are the next steps.

> I just hopped across eight time zones

Have fun at Scala Days!

rkuhn commented:
Thinking this through we came to the conclusion that you are right: state changes are related to how incoming commands are handled and should not be conflated with domain events (which change the state data). On the other hand, modeling something as PersistentFSM means that the state that it was in does usually have some relevance in the domain model. The conclusion is that state changes should be emitted as events by themselves and domain events should also be emitted one-by-one by themselves to keep them fully accessible to other consumers of the event log. The ordering of these two different kinds of events allows some freedom—obviously everything related to one transition needs to stick together—because during recovery no new commands are being processed.

Thanks for your inputs, this is a great PR/discussion!

leonidb commented:
@rkuhn @patriknw 

>  The conclusion is that state changes should be emitted as events by themselves and domain events should also be emitted one-by-one by themselves to keep them fully accessible to other consumers of the event log

Other customers of the event log will also receive the state change events, as all events are persisted with the same `persistenceId`, so they will have to ignore those events, right?

> obviously everything related to one transition needs to stick together—because during recovery no new commands are being processed

Is it good enough to persist such events with `def persist[A](events: immutable.Seq[A])...` from `Eventsourced`?

rkuhn commented:
> Other customers of the event log will also receive the state change events, as all events are persisted with the same persistenceId, so they will have to ignore those events, right?

Yes.

> Is it good enough to persist such events with `def persist[A](events: immutable.Seq[A])...` from Eventsourced?

Yes, exactly.

patriknw commented:
+1

leonidb commented:
@patriknw @rkuhn 
Apologies for the delay, a busy week, and the separation of State Change events and Domain events was harder than I thought. 
This commit has some flaky parts, hope that the general direction is good.
Waiting for your feedback.

rkuhn commented:
commented on the last commit: direction looks good, thanks!

patriknw commented:
@leonidb the direction is looking good

rkuhn commented:
This is looking really good! I just reviewed the whole PR again and I think there is not much we can do about the duplication with the `akka.actor.FSM`—unless someone has a brilliant idea that works within the desire to keep binary compatibility. So what is missing now is documentation (which includes Java API test coverage).

leonidb commented:
@rkuhn 
>So what is missing now is documentation (which includes Java API test coverage).

I'll start by implementing the existing Scala tests in Java

>there is not much we can do about the duplication with the akka.actor.FSM

So, should it stay duplicated, or somehow get merged into the original `akka.actor.FSM`?

rkuhn commented:
It will need to stay duplicated, but of course it cannot be in a package named `temp` ;-) I’d propose to move all the pieces of this PR into a new package named `akka.persistence.fsm`, that should make it clear what is going on.

leonidb commented:
OK, I'll start with that and then do the documentation/tests.

Are there any time limitations/release dates I should be aware of?

rkuhn commented:
for the `master` branch you’ll have at least two more weeks of “solitude” while we finish Akka Streams 1.0-RC1; afterwards there will be some activity to prepare master for 2.4-M1; so I’d say now is the time to do it :-)

patriknw commented:
Extracting the common things to an internal `FSMLike` trait would not break binary compatibility of existing `FSM`, or would it?

rkuhn commented:
Potentially doable, but probably a lot of work and tricky verification and hard thinking; if someone wants to do it then go ahead, otherwise I wouldn’t it call required at this point. The problems I see revolve around the case class nature of StateData.

leonidb commented:
I'd love to give it a try later, if I have enough time.

patriknw commented:
It can be done as step 2, i.e. not in this PR.
Thanks for your excellent work on this one @leonidb!

leonidb commented:
OK, thanks!

leonidb commented:
@rkuhn @patriknw
I've started to work on documentation, where do you think should I add it? 
Can you recommend me what should it include and how should it be structured?

Is there anything left besides the documentation, before the PR can be merged?

ktoso commented:
Awesome! We talked about plans concerning persistence over the last days and this PR as well - it only needs documentation and can be merged we thought :-)

The docs can just be added as a section in the `java/persistence.rst`, `scala/persistence.rst` and `lambda-persistence.rst` files.

leonidb commented:
@ktoso 
Not sure about `java/persistence.rst` since Java implementation of Persistent FSM - `AbstractPersistentFsmActor` uses Java 8 lambdas. If I understand correctly, `java/persistence.rst` shouldn't include the usage of Java 8 lambdas, right?

ktoso commented:
You're right - just `persistence-lambda`, and we'll move all lambda docs to the `java/persistence` file soon - because Akka 2.4 will require Java 8+ the non-lambda docs will not be needed for this release :-)

rkuhn commented:
Awesomely done, @leonidb! Now the only thing missing before clicking the green button is to make that button green again ;-) (which means: please squash into one commit and the rebase onto current master)

leonidb commented:
@rkuhn 
I haven't commited the Java doc yet, and Scala doc probably needs some fixes. 
Can this be done after merging the PR?

drewhk commented:
> I haven't commited the Java doc yet, and Scala doc probably needs some fixes. 
Can this be done after merging the PR?

please do it in this PR, you will need a rebase anyway. Nice work btw!

leonidb commented:
I'm working on squashing and rebasing with Git and I need some advice.

Here is what I did so far:
I've squashed all commits to one and pushed this unified commit to my fork.
Locally I've rebased the commit onto the current master and resolved the conflicts. 
I also had to add some small changes to tests.

Now, if I'm not mistaken, I need to merge all that into leonidb/akka repository.
What's the recommended way to do it - Rebase, or maybe create a new fork/pull request and then only rebase the unified commit on top of it?

Thanks,
Leonid

patriknw commented:
If you have one commit that it rebased onto master you should be able to force push (git push -f) that to the same branch as before to update this pull request.

leonidb commented:
@patriknw  thanks
I've pushed 2 commits - one is the unified commit of all previous commits, the second contains the fixes I've made to make the tests pass.

There are some issues still left to solve:
1. Compilation fails because I've used ```Consumer``` functional interface from Java 8. Can be solved by using a different interface with the same signature.
2. I've set up the test using ```PersistenceSpec.config("leveldb", "PersistentFSMActorSpec", "off")```, couldn't make it pass with the default ```serialization: String = "on"``` value. Is this something that should be fixed? If so, how?
3. Is there a more elegant way to set up the Java test than:
```Scala
private static Option<String> none = Option.none();
    @ClassRule
    public static AkkaJUnitActorSystemResource actorSystemResource =
            new AkkaJUnitActorSystemResource("PersistentFSMJavaTest", PersistenceSpec$.MODULE$.config(
                    "leveldb", "AbstractPersistentFsmActorTest", "off", none.asScala()));
```
Is there anything else I'm missing?
Once those issues are resolved, I'll squash the commits into one, with a better description.

patriknw commented:
@leonidb as of today master is on jdk8, so you can use the Consumer interface, otherwise I think there is something similar in akka.japi

patriknw commented:
@leonidb we should find a better way for `PersistenceSpec$.MODULE$.config`.
Is it possible to change PersistenceSpec from a trait to an abstract class? Then you will get static forwarder for `PersistenceSpec.config`. Otherwise we can rename PersistenceSpec "companion" to `PersistenceTestConfig` or something

patriknw commented:
@leonidb what is the error when serialize = "on", some message that is not serializable? If it is only intended for local message sends you can add the marker `NoSerializationVerificationNeeded` to the message.

sirocchj commented:
@patriknw I have a question on this PR: is this going to be backported to the 2.3.x series?

patriknw commented:
We normally don't backport new features to 2.3.x. You will see milestone releases of 2.4 soon and 2.4 final is not that far out.

sirocchj commented:
@patriknw thanks for the reply. I based the question on the fact that 2.4 is going to be targeting JDK8 runtime

rkuhn commented:
@sirocchj It would be a very deep rabbit hole if we started to back-port features to support them on Java 6, the idea of the open-source development of Akka is to track the current state of the art.

If on the other hand you want a supported version of these features on a discontinued version of Java then Typesafe could provide that as a commercial offering, ping me (roland AT typesafe.com) for details if you are interested.

leonidb commented:
@patriknw 
> as of today master is on jdk8, so you can use the Consumer interface

Well, currently I see a compilation error in Jenkins build output:
``` [error] /mnt/akka/jenkinsakka/localhome-a4/workspace/pr-validator-per-commit-jenkins/akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala:130: not found: type Consumer```

> there is something similar in akka.japi

I can use ```UnitApply<I>``` instead, however ```AbstractPersistentFsmActorTest``` test uses lambdas and it needs to be compiled with jdk8.

leonidb commented:
@patriknw 
> Is it possible to change PersistenceSpec from a trait to an abstract class

```PersistentSpec``` is used as a mix in with ```AkkaSpec```, which is an abstract class, so it's not straight forward. Might be possible with a bigger code refactoring, probably not worth the effort.

> Otherwise we can rename PersistenceSpec "companion" to ```PersistenceTestConfig```

This works, I can also rename ```config``` function to ```apply``` in this case, to make it look better in Scala. This rename slightly affects 12 files, but it does look harmless.

Tell me if you think it's worth it.

sirocchj commented:
@rkuhn thanks for elaborating!
I sure do understand your commitment and decision to support the current version of the JDK and not back-port new features on older versions.
Being currently stuck with 2.3.x until migration to 8 is completed on our side (yeah, I know...!!), I was just wondering on this one in particular as I am currently prototyping an eventsourced FSM (and was hoping to do extends PersistentActor with FSM). I guess I'll prototype with 2.4.M1 on 8 and wait for final before rolling out. I am sure you agree makes this PR will make life easier to many, compared to becomes.
A final note, I believe the docs on persistence, particularly the brief section on state machines, would need some tlc when you get a chance.

patriknw commented:
@leonidb you could try `abstract PersistenceSpec extends AkkaSpec`, but otherwise the rename is fine

Wasn't the failure before we switched to running pr-validation with jdk8? I will trigger a new build.

patriknw commented:
PLS BUILD

leonidb commented:
@patriknw 
> you could try abstract PersistenceSpec extends AkkaSpec

That's what I did eventually

> Wasn't the failure before we switched to running pr-validation with jdk8

leonidb commented:
@patriknw 
> what is the error when serialize = "on", some message that is not serializable?

Besides messages being not serializable, there is an issue with actor creation, I get this exception:
```java.lang.IllegalArgumentException: pre-creation serialization check failed at [akka://PersistenceSpec/user/$a]...```

So I've tried to change 
```Scala
system.actorOf(Props(new WebStoreCustomerFSMActor(persistenceId, dummyReportActorRef)))
``` 
into 
```Scala 
system.actorOf(Props(classOf[WebStoreCustomerFSMActor], persistenceId, dummyReportActorRef))
```
which led me to a new exception:
```java.lang.IllegalArgumentException: no matching constructor found on class akka.persistence.fsm.PersistentFSMActorSpec$WebStoreCustomerFSMActor for arguments [class java.lang.String, class akka.actor.RepointableActorRef]```
Explicit addition of domain event ```ClassTag```  solves the issue and makes the test pass:
```Scala
system.actorOf(Props(classOf[WebStoreCustomerFSMActor], persistenceId, dummyReportActorRef, ClassTag(classOf[PersistentFSMActorSpec.DomainEvent])))
```
This has surprised me, as I expected ```ClassTag``` to work implicitly. It does work as expected in Java:
```Java
system.actorOf(Props.create(WebStoreCustomerFSMActor.class, persistenceId, dummyReportActorRef));
```
Am I missing something here, regarding the usage of ```ClassTag```?

rkuhn commented:
@leonidb Implicit arguments (like the ClassTag in your case) still need to be passed explicitly at the JVM level, they are normal method arguments. The only thing that is special about them is on the Scala level where the compiler can automatically find the appropriate values and fill them in. When you invoke a constructor using reflection (which is what Props does here) the compiler has no clue what you are doing and therefore cannot know that one of the arguments should be implicit.

leonidb commented:
@rkuhn
So I'll probably go with
'''Scala
system.actorOf(Props.create(WebStoreCustomerFSMActor.class, persistenceId, dummyReportActorRef));
'''
And turn on the serialization check, right?

leonidb commented:
@rkuhn 
I've notices that 2.4-M1 was released yesterday, does it mean any changes regarding this PR? Can it be merged?
Does it somehow affect the merge process? Should I still squash the changes, including the latest test fixes, into one commit, rebase on akka/master and ```push -f``` to my branch?

rkuhn commented:
Yes, 2.4-M1 does not change anything here: that release is just a snapshot of the master branch that has named artifacts on Maven Central.

Concerning the Props, our recommended pattern (that we should therefore also follow in the documentation) is to make a method on the companion object that constructs the Props, and in this method you use the `new ...` syntax to get all the compiler checking and implicit argument resolution. These Props will be serializable because the created factory object lives in the static scope of the object (right next to the Actor’s declaration in this case). Then in the test you just use `PersistentFSMActorSpec.webstoreProps(...)` to obtain the Props to be passed to `system.actorOf`.

leonidb commented:
@rkuhn 
Thanks!
I'll fix the props later today.

leonidb commented:
I haven't closed the PR, I guess it was closed because I've used ```git push -f``` incorrectly.
I've reopened it. Hope that everything is correct this time.

leonidb commented:
PLS BUILD

rkuhn commented:
OK TO TEST

leonidb commented:
@ktoso @patriknw 
I've looked at Jenkins build job log and it looks like the build fails when building the Javadoc.
When I run the build locally I also get the same error:
```[error] (akka-persistence-experimental/genjavadoc:doc) javadoc returned nonzero exit code```
I see documentation related errors in SBT output, however many of them are not related to persistence module, so they are probably not an indication of a problem.
Can you have a look? 
How can I investigate it further?

ktoso commented:
The error is:

```
ktoso @ 月~/code/akka [leonidb-master*]
$ cat ex | perl find-error.pl
[error] /Users/ktoso/code/akka/akka-persistence/target/java/akka/persistence/fsm/AbstractFSM.java:295: error: invalid use of @return
[error]    * @return current state descriptor
[error]      ^
[error] /Users/ktoso/code/akka/akka-persistence/target/java/akka/persistence/fsm/FSM.java:448: error: bad use of '>'
[error]    * If you want to trigger an event like <code>S -> S</code> for {@link #onTransition} to handle use {@link #goto} instead.
[error]                                                    ^
```

patriknw commented:
Thanks Konrad, I'm smiling every time I'm reminded that we have to use perl to parse the error log

ktoso commented:
Yeah, lol ;-)

patriknw commented:
If not too much work it would be cool to run the script as an extra build step in jenkins, so that we get the filtered errors in the log.

ktoso commented:
Yes agree, we should think of something to improve dev experience here, its too confusing for contributors and ourselves alike. Ticketified https://github.com/akka/akka/issues/17611

leonidb commented:
@ktoso
Thanks, that was really helpful! All those errors are indeed confusing.
I've committed the fix, hope it works this time

ktoso commented:
this time:
```
[error] /mnt/akka/jenkinsakka/localhome-a0/workspace/pr-validator-per-commit-jenkins/akka-persistence/target/java/akka/persistence/fsm/PersistentFsmActor.java:44: error: reference not found
[error]    * @return {@link ClassTag} of domain event base type
[error]
```

leonidb commented:
@ktoso 
Thanks! 
For some reason now I don't get those errors when running locally. 
I did manage to run the perl script on Jenkins output, but missed this one error.

leonidb commented:
@ktoso @patriknw @rkuhn 
So now, if everything is OK with the Java Docs fixes, I need to squash the latest commits again, right?
Is there anything else left to do before that?

rkuhn commented:
No, just squash and force-push and we’ll click the button :-)

patriknw commented:
that is awesome

leonidb commented:
Squashed and pushed

patriknw commented:
@leonidb Merged! Yay! Thanks for an awesome contribution.

patriknw commented:
Refs #15279

ktoso commented:
Awesome! Thanks so much @leonidb for pulling this through!

leonidb commented:
Great to see it merged, finally :)

Thanks, @patriknw, @ktoso, @rkuhn,
I've learned a lot while working on this feature, both while coding and from your feedbacks!

hepin1989 commented:
will take a deep look into it,thanks

sovaalexandr commented on akka-persistence/src/main/scala/akka/persistence/fsm/FSM.scala:
```diff
@@ -0,0 +1,1179 @@
+      }
+      currentState = nextState
+      val timeout = if (currentState.timeout.isDefined) currentState.timeout else stateTimeouts(currentState.stateName)
+      if (timeout.isDefined) {
```
@leonidb Is it for performance consideration you unwrap value from Option here?
